Small optimizations to AiiDA workflow implementation

sphuber opened the pull request:
The workflow is simplified and improved in two ways

 * The final output PDF is already stored as a PDF in AiiDA's database
   so there is no need to pass it through a PDF library again. It suffices to
   copy the file as bytes to a file in the current working directory.
* The step that prepares the latex macros extracts data from the output
  produced by the gmsh and fenics programs. By using a normal function,
  the provenance of the extracted numbers is "lost". Of course from the main
  script it is still obvious where they come from, but when just looking at the
  provenance graph, it is not evident where the numbers of `--domain-size`
  and `--num-dofs` come from. By using a `calcfunction`, this provenance is
  explicitly kept as well.

sphuber commented:
Note that I haven't actually tested running this, so there may be subtle bugs in the changes, but the gist of it should work. Let me know if you get the chance to run it and run into problems, I will quickly provide feedback on how to correct it.

sphuber commented:
One more thing I forgot to mention in my email reply: it might be nice and instructive to show the final generated provenance graph. You can do so by running `verdi node graph generate PK` where you replace `PK` with the value returned by `paper["paper_pdf"].pk`. Might make for a nice figure for the paper.

pdiercks reviewed (approved):
Hi @sphuber , 
Thanks for the PR and your email reply! 

With the suggested changes everything works as planned. 
IMO for the calcfunctions it would make sense to return Float or Int, and do the conversion to string when passing to the `ShellJob`.
However, the calcfunctions are not shown in the provenance graph. I guess this would not be the case if the Float/Int is passed to the `ShellJob` directly? Or is it rather because I am not using a `WorkChain`?

I used the command you suggested and this is the [provenance graph](https://bambund-my.sharepoint.com/:b:/g/personal/philipp_diercks_bam_de/ESN3BMwaQL1DhdJ672PVxIoBAbaK1P21k4pUbmbPOHaRfg?e=jttBxK) that I got

sphuber commented on simple_use_case/aiida/suc_v2b1.py:
```diff
@@ -63,17 +63,18 @@
+    stdout = fenics_stdout.get_content()
     ndofs = stdout.split("Number of dofs used:")[1]
-    return "".join(ndofs.split())
+    return Str("".join(ndofs.split()))
```
If it is an integer, it would indeed make sense to use `Int`. I simply used `Str` since I wasn't sure that the type would be a single integer.

sphuber commented:
> IMO for the calcfunctions it would make sense to return Float or Int, and do the conversion to string when passing to the `ShellJob`.
> However, the calcfunctions are not shown in the provenance graph. I guess this would not be the case if the Float/Int is passed to the `ShellJob` directly? Or is it rather because I am not using a `WorkChain`?

You are absolutely right, the reason they don't show up is because they are not actual inputs to the shell job, but they are converted to normal strings using `str(.value)`. I forgot that the `arguments` keyword of the `launch_shell_job` expects a `List` node, which cannot contain any nodes itself. So you are right that currently you are forced to convert the nodes, and so this does not have the effect I intended as that will lose the explicit link between the node and the next process in the graph.

Conceptually it might still be possible but it would require a change in `aiida-shell` first. When I was typing the response, I was thinking that it might make sense to pass more than just `SinglefileData` nodes as inputs that will be expanded, but the current implementation doesn't allow for this.

The basic node types, for example, like `Float`, `Int` and `Str` could be perfectly passed as an input and the `ShellJob` would then convert them to a `str` so it can be added as a command line argument. I will work a bit on the `aiida-shell` side what this would look like. The main question is whether we can allow "any" `Node` type. The `ShellJob` would need to call a default method that will convert the node's content to a str value. But this will of course only make sense for certain node types.

sphuber commented:
Thanks for the review @pdiercks . I have implemented the required functionality in `aiida-shell` to pass more than just `SinglefileData` nodes as inputs. I have just release `v0.2.0` and updated this PR to make use of that version. It would be great if you can try to update and run the new implementation and then check the new provenance graph.

pdiercks commented:
Hi @sphuber ,
thanks for updating the workflow and the documentation to make use of `aiida-shell` version `v0.2.0`. I just ran the script and everything works fine. Here is the updated [provenance graph](https://bambund-my.sharepoint.com/:b:/g/personal/philipp_diercks_bam_de/EeOv27ILyaxGinkJ4gIhfF4BP7xl7a1D4Ax9En73bfbqnA?e=0GR7e6) which now also shows the calcfunctions and the float and integer values for `domain_size` and `num_dofs` respectively. Very cool!

I just renamed the file since the previous name didn't make any sense anymore. 
From my side this is ready to merge.
